nebu/bakedpdf.py

```diff
@@ -0,0 +1,129 @@
+                                        tag=tags, remove_blank_text=True)
+
+        root_adapter = build_adapters_tree(collxml_etree)
+
+        binder = root_adapter.to_model()
+        return binder
+
+    def build_adapters_tree(collxml_etree):
+        """Binders require bottom-up construction (sub-binders are passed into
+        the constructor). So we'll (1) parse the element tree top-down and
+        store the needed information in Adapter nodes and then (2) use
+        post-order traversalon the Adapter tree to construct Binders from their
+        sub-binders.
+        """
+        node = None  # the node currently being parsed
+
+        for event, element in collxml_etree:
+            tag = element.tag.split('}')[-1]  # tag without namespace
+
+            if tag == 'title':
+                assert node  # adapter w/ 'collection' tag should already exist
+                node.title = element.text
+
+            elif event == 'start':
+                new_ = Adapter(tag, element.attrib)
+                if node:
+                    node.add_child(new_)
+                node = new_
+
+            elif event == 'end':
+                node = node.parent
+
+        # After the last `end` event is processed, `node` is the root node
+        root_node = node
+        return root_node
+
+    class Adapter(object):
+        def __init__(self, tag, attrib, title=''):
+            self.tag = tag
+            self.attrib = dict(attrib).copy()
+            self.title = title
+            self.parent = self
+            self.children = []
+
+        def add_child(self, child):
+            child.parent = self
+            self.children.append(child)
+            return self
+
+        def __iter__(self):
+            return iter(self.children)
+
+        def iter(self, tag='*'):
+            if tag == '*' or self.tag == tag:
+                yield self
+
+            for child in self:
+                yield from child.iter(tag)
+
+        def to_model(self):
+            ch_models = [child.to_model() for child in self.children]
+            meta = {'title': self.title}
+
+            if self.tag == 'collection':
+                return Binder(self.title, nodes=ch_models, metadata=meta)
+            elif self.tag == 'subcollection':
+                return TranslucentBinder(nodes=ch_models, metadata=meta)
+            elif self.tag == 'module':
+                module_id = self.attrib.get('document')
+                data = find_doc_data(module_id)
+                meta['license_url'] = None  # required by `Document`
+                return Document(module_id, data, metadata=meta)
+            else:
+                raise Exception('unrecognized tag: {}'.format(self.tag))
+
+    def find_doc_data(module_id):
+        """Locates and reads a cnxml file/module with given `module_id`
+        """
+        try:
+            mod_loc = source_dir / module_id / 'index.cnxml.html'
```

Comment:
Per our discussion in person, I am going to use `cnx-litezip` to handle the module parsing part and in my code, I will "just" call `Module.id` and `Module.data`. Also, I will expand litezip to look for modules with the filenames `index.cnxml.hmtl` and `index.xhtml`, somehow, as you suggested.

Reply:
Great, and thanks for documenting, especially since it brought to light my failure to communicate. I don't think we'll need changes to litezip per se - it depends on there being a index.cnxml file to determine the existence of a module directory, and all our ways of creating module-directories do that (complete.zip, neb get --whatever). I was just suggesting that when assembling the single-html file, go ahead and use an index.xhtml file, in the absence of the index.cnxml.html file.